File: 2023/BM/32/codes/gate.c
#include <stdio.h>
#include <math.h>

// Function to calculate |Z|
double calculate_Z_magnitude(double omega_0) {
    return 1e3 * (sqrt(100 + pow(omega_0, 2)) / omega_0);
}

int main() {
    // Open the file for writing
    FILE *file = fopen("data.dat", "w");
    
    if (file == NULL) {
        printf("Error opening file!\n");
        return 1;
    }

    // Generate values for omega_0 from 1 to 25
    for (double omega_0 = 1.0; omega_0 <= 25.0; omega_0 += 0.01) {
        // Calculate corresponding |Z| values
        double Z_magnitude = calculate_Z_magnitude(omega_0);
        
        // Print values to the file
        fprintf(file, "%.2f %.4f\n", omega_0, Z_magnitude);
    }

    // Close the file
    fclose(file);

    return 0;
}

File: 2023/BM/32/codes/gate.py
import matplotlib.pyplot as plt

# Read data from data.dat
data = []
with open('data.dat', 'r') as file:
        for line in file:
                    omega_0, Z_magnitude = map(float, line.split())
                            data.append((omega_0, Z_magnitude))

                            # Separate data into omega_0 and Z_magnitude
                            omega_0_values, Z_magnitude_values = zip(*data)

                            # Plot the values
                            plt.plot(omega_0_values, Z_magnitude_values, label=r'$|\mathbf{Z}| = 10^3 \left(\frac{\sqrt{100 + \omega_0^2}}{\omega_0}\right)$')

                            # Highlight the point at omega_0 = 10 in a different color
                            highlight_omega_0 = 10
                            highlight_Z_magnitude = Z_magnitude_values[omega_0_values.index(highlight_omega_0)]
                            plt.scatter(highlight_omega_0, highlight_Z_magnitude, color='red', label=f'Highlight at $\omega_0 = {highlight_omega_0}$')

                            # Add labels and legend
                            plt.xlabel(r'$\omega_0$')
                            plt.ylabel(r'$|\mathbf{Z}|$')
                            plt.legend()

                            # Show the plot
                            plt.grid(True)
                            plt.show()
